department-of-veterans-affairs/va.gov-cms
# VACMS 3450: Repair improperly-parented paragraphs resulting from entity_clone operations.

## Description

See #3450. 

Briefly, the `entity_clone` form provided until recently (see #3449) the ability to uncheck the checkboxes directing whether to clone paragraphs entities.  The result was a large number of nodes that all had references to different revisions of the same paragraph entities.

This PR addresses this problem by:
1. Querying for a list of all paragraph reference fields.
1. For each of those paragraph reference fields on a node content type, query for a list of all paragraphs with revisions that point to distinct parent nodes.
1. For each of those paragraphs, create a new clone that is parented to the proper node and save both the clone paragraph and the parent node.

This script should be idempotent; repeated runs should make no changes and, from my testing, behave as intended.

## Testing done
Executed the script repeatedly in the course of development, refreshing the database from production as needed.  Verified that new revisions are created on the nodes in question, and that each revision corresponds with the creation and attachment of a new paragraph entity with the content of the old paragraph.

## Screenshots
Consider a paragraph with id `8378`.  

The following is a greatly abbreviated list of revisions for this paragraph, showing that it is associated with many different nodes via the `parent_id` field:
<img width="1033" alt="Screen Shot 2020-11-17 at 3 03 32 PM" src="https://user-images.githubusercontent.com/1318579/99450287-2f888f80-28e6-11eb-84fd-ff181ea89c1b.png">

Before running the script, we can look at the paragraphs associated with one of these nodes, the node with id `8067`:
<img width="1033" alt="Screen Shot 2020-11-17 at 3 02 16 PM" src="https://user-images.githubusercontent.com/1318579/99450533-85f5ce00-28e6-11eb-99b5-a536f3efb1d6.png">

After running the script, we should be able to look at the node with id `8067` and see new paragraph revisions associated:
<img width="1038" alt="Screen Shot 2020-11-17 at 3 10 38 PM" src="https://user-images.githubusercontent.com/1318579/99450968-1f24e480-28e7-11eb-8781-6a9011d3e4be.png">

And we should be able to look at the node and see three new revisions in the UI:
<img width="1034" alt="Screen Shot 2020-11-17 at 3 13 30 PM" src="https://user-images.githubusercontent.com/1318579/99451196-76c35000-28e7-11eb-917c-141e3c5b9973.png">

And comparing the last previously existing revision and the latest new revision, we should see that new paragraphs were added to the node with content corresponding to that in the existing paragraphs:
<img width="637" alt="Screen Shot 2020-11-17 at 3 14 08 PM" src="https://user-images.githubusercontent.com/1318579/99451288-9fe3e080-28e7-11eb-898f-f2f7d85033da.png">

If, however, we consult the original node that `8067` improperly cloned, `2803`, we should see no added revisions either in the database:
<img width="1033" alt="Screen Shot 2020-11-17 at 3 16 58 PM" src="https://user-images.githubusercontent.com/1318579/99451590-0832c200-28e8-11eb-8603-1b79b8c672ef.png">

or in the UI:
<img width="1015" alt="Screen Shot 2020-11-17 at 3 17 54 PM" src="https://user-images.githubusercontent.com/1318579/99451627-1385ed80-28e8-11eb-8bc5-a71300a4c12e.png">

And its content should remain intact:
<img width="1029" alt="Screen Shot 2020-11-17 at 3 18 29 PM" src="https://user-images.githubusercontent.com/1318579/99451676-28fb1780-28e8-11eb-98fa-268126d208d8.png">

## QA steps
1. Query for revisions of a paragraph entity and verify that it has multiple parents:
  - e.g. `lando drush sqlq 'SELECT DISTINCT id,parent_id FROM paragraphs_item_revision_field_data WHERE id = 8378;'`
1. Query for revisions of all paragraph entities associated with a specific parent:
  - e.g. `lando drush sqlq 'SELECT DISTINCT id,parent_id FROM paragraphs_item_revision_field_data WHERE parent_id = 8370;'`
1. Run the script.
  - e.g. `lando drush scr ./scripts/content/VACMS-3450-paragraphs-should-have-a-single-parent.php --no-halt-on-error`
  - the `--no-halt-on-error` seemed to be needed for me because of an issue elsewhere
1. Query for revisions of all paragraph entities associated with a specific parent:
  - e.g. `lando drush sqlq 'SELECT DISTINCT id,parent_id FROM paragraphs_item_revision_field_data WHERE parent_id = 8370;'`
  - [ ] verify that new paragraph revisions exist pointing to this parent

### Definition of Done

### Select Team for PR review

### Does this PR need review from a Product Owner

### CMS user-facing annoucement

Is an announcement needed to let editors know of this change?
